src/mmds/src/lib.rs

```diff
@@ -174,14 +256,123 @@ fn respond_to_get_request(request: Request) -> Response {
                 StatusCode::PayloadTooLarge,
                 Body::new(e.to_string()),
             ),
-            MmdsError::NotInitialized => unreachable!(),
+            _ => unreachable!(),
         },
     }
 }
 
+fn respond_to_put_request(request: Request, token_headers: TokenHeaders) -> Response {
+    // Reject `PUT` requests that contain `X-Forwarded-For` header.
+    if request
+        .headers
+        .custom_entries()
+        .contains_key(REJECTED_HEADER)
+    {
+        let error_msg = RequestError::HeaderError(HttpHeaderError::UnsupportedName(
+            REJECTED_HEADER.to_string(),
+        ))
+        .to_string();
+        return build_response(
+            request.http_version(),
+            StatusCode::BadRequest,
+            Body::new(error_msg),
+        );
+    }
+
+    let uri = request.uri().get_abs_path();
+    // Sanitize the URI into a strict json path.
+    let json_pointer = sanitize_uri(uri.to_string());
+
+    // Only accept PUT requests towards TOKEN_PATH.
+    if json_pointer != PATH_TO_TOKEN {
+        let error_msg = format!("Resource not found: {}.", uri);
+        return build_response(
+            request.http_version(),
+            StatusCode::NotFound,
+            Body::new(error_msg),
+        );
+    }
+
+    // Get token lifetime value.
+    let ttl_seconds = match token_headers.x_metadata_token_ttl_seconds() {
+        Some(ttl_seconds) => ttl_seconds,
+        None => {
+            return build_response(
+                request.http_version(),
+                StatusCode::BadRequest,
+                Body::new(NO_TTL_PROVIDED_ERROR),
+            );
+        }
+    };
+
+    // Generate token.
+    let response = match MMDS.as_ref() {
+        Some(mmds) => mmds
+            .lock()
+            .expect("Poisoned lock")
+            .generate_token(ttl_seconds),
+        // This path is unreachable because MMDS entity has been previously
+        // checked.
+        None => unreachable!(),
+    };
+    match response {
+        Ok(token) => build_response(request.http_version(), StatusCode::OK, Body::new(token)),
```

Comment:
Indeed there is no JSON logic in the code, the `PUT` request does return only a String. This is consistent to EC2's IMDSv2 behavior and I personally believe it's easier to operate with, since one could have something like:
```bash
TOKEN=curl -X PUT "http://169.254.169.254/latest/api/token" -H "X-metadata-token-ttl-seconds: 21600"
curl -H "X-metadata-token: $TOKEN" -v http://169.254.169.254/latest/meta-data/
```

Reply:
In this case, we need to make sure we set the `Content-Type` header in the response to `text/plain`.